File: 150 Evaluate Reverse Polish Notation/150.evaluate-reverse-polish-notation.py
#
# @lc app=leetcode id=150 lang=python3
#
# [150] Evaluate Reverse Polish Notation
#

# @lc code=start
from typing import List

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        # 1. Initialize an empty stack to store operands
        stack = []

        # 2. Define operations in a dictionary for easy access
        operations = {
            '+': lambda x, y: x + y,
            '-': lambda x, y: x - y,
            '*': lambda x, y: x * y,
            '/': lambda x, y: int(x / y),  # Python automatically performs floor division for int
        }

        # 3. Iterate over each token in the tokens list
        for token in tokens:
            if token in operations:
                # 4. If token is an operator, pop the last two numbers from stack
                y, x = stack.pop(), stack.pop()
                # 5. Apply the operation and push the result back onto the stack
                stack.append(operations[token](x, y))
            else:
                # 6. If token is a number, convert it to integer and push onto stack
                stack.append(int(token))
        
        # 7. Final result will be the only element left in the stack
        return stack[0] 
# @lc code=end

File: 150 Evaluate Reverse Polish Notation/150.evaluate-reverse-polish-notation.cpp
/*
 * @lc app=leetcode id=150 lang=cpp
 *
 * [150] Evaluate Reverse Polish Notation
 */

// @lc code=start
#include <vector>
#include <string>
#include <stack>
#include <stdexcept>
using namespace std;

class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        // Step 1: Initialize a stack for intermediate results.
        stack<int> stack;

        // Step 2: Iterate through tokens to evaluate expression.
        for (const string& token : tokens) {
            if (token == "+" || token == "-" || token == "*" || token == "/") {
                // Step 3: Pop two elements to apply operation.
                int b = stack.top(); stack.pop();
                int a = stack.top(); stack.pop();

                // Step 4: Perform operation based on token.
                if (token == "+") stack.push(a + b);
                else if (token == "-") stack.push(a - b);
                else if (token == "*") stack.push(a * b);
                else if (token == "/") stack.push(a / b); // Integer division.
            } else {
                // Step 5: Convert number token to integer and push to stack.
                stack.push(stoi(token));
            }
        }

        // Step 6: Return the result from the stack.
        return stack.top();
    }
};

// @lc code=end

